servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/IoThreadFactory.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.transport.api;
+
+import java.util.concurrent.ThreadFactory;
+
+/**
+ * Thread factory for use with {@link IoExecutor}.
+ *
+ * @param <T> Type of threads created
+ */
+public interface IoThreadFactory<T extends Thread & IoThreadFactory.IoThread> extends ThreadFactory {
```

Comment:
> * …
> * …
> * Not necessary to change now, just thinking out loud: because we have a way to enforce  what else a `Thread` needs to implement, we can also enforce `AsyncContextMapHolder` interface to make sure users don't see perf degragation if they have a custom `ThreadFactory`. Would be glad to hear @Scottmitch opinion on that. The only hesitation I have: if we add it later, it will be API breaking change. Maybe we should defer cherry-picking this PR into 0.41 branch until after we have 0.41.4 release to give us some time to discuss.

Since `AsyncContextMapHolder` is a public API type it does not seem unreasonable to add to the required interface types for `IoThread`. For implementation of the interface we will have to provide sample code as the interface cannot be implemented via default methods (needs state).